lib/std/smartptrs.nim

```diff
@@ -0,0 +1,212 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
```

Comment:
In C++, an important feature is to allow custom deleter functions (https://en.cppreference.com/w/cpp/memory/unique_ptr) so that the unique_ptr can call the right "free" function - this allows them to be used with custom allocators, to manage file handles and other resources that require "manual"  disposal etc.

In particular, limiting the type to the shared allocator would be an unfortunate design, becuse typically when you're doing manual resource management, you're doing so because you have special requirements.

Reply:
now it supports deleter functions

Reply:
Please add a description of smart pointers vs Nim references, and under what situations this module would be useful. For a rough example, see Python's [weakref](https://docs.python.org/3/library/weakref.html) module.